### PR TITLE
cardano-wallet server on _any_ arbitrary testnet

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

N/A

# Overview

- e0f5ef5aea50d01da290e53c795da4572303171c
  fix mainnet genesis hash
  
- dc0048d8536d0000bf5891512b2dd68270a625e6
  embed network magic inside Testnet discriminant
  The rationale is that, for mainnet, it is both known and not actually
required (addresses don't have protocol magic attributes). But for testnets
(plural), we may want an easy way to switch from a network to another.

- 22b914ff71331a40ed0fd9e644d976757f45ee7b
  allow passing custom testnets configuration via command-line
    e.g.

  ```
  $ cardano-wallet serve --mainnet \
    --node-socket ../cardano-node/state-node-mainnet/node.socket
  ```

  ```
  $ cardano-wallet serve --testnet 1097911063 \
    --genesis /nix/store/7106qjwral616j6p5049m7hhi5ih0r33-testnet-genesis.json \
    --node-socket ../cardano-node/state-node-testnet/node.socket
  ```

- 70cd8be7fe5875e65f9a86b6651d6bcacce41a32
  fixup fromGenesisTxOut
  
- 7fc880d51a1eb0f0f87cf83ca3a55472022f94c9
  adjust NetworkDiscriminant in test code to cope with recent change
  
- 7b9727a1d538c9089b18ec63a7411ce132901073
  remove hard-coded references to network discriminant from integration scenarios
  And rely instead on class constraints. There's no reason to specialize the network discriminant too early. The only moment
where we would want to do that is when calling 'serveWallet'.

- 896bf0f134ea002ec7fe0fd1e94c0ee944388bcc
  re-generate nix machinery
  
- ab155f7ae9a663d941291efa0b50c3479a1d8083
  improve error message when failing to parse genesis configuration
  

# Comments

This last commit is really not interesting, sorry for this. It is simply adjusting the network discriminant in many places. Note that I have switched from testnet to mainnet in a few places because it's easier to use and doesn't quite matter most of the time.

### CODE REVIEW

**Comment by KtorZ** [comment]
bors try

File: lib/byron/exe/cardano-wallet-byron.hs [hunk 1]
```diff
@@ -176,37 +192,46 @@ cmdServe = command "serve" $ info (helper <*> helper' <*> cmd) $ mempty
...
-        let networkParams = knownParameters proxy
+        (discriminant, bp, vData, block0) <- case networkConfig of
+            MainnetConfig (discriminant, vData) -> pure
+                ( discriminant
+                , blockchainParameters @'Mainnet
+                , vData
+                , emptyGenesis (blockchainParameters @'Mainnet)
+                )
+            TestnetConfig (discriminant, vData) genesisFile -> do
+                (genesisData, genesisHash) <- unsafeRunExceptT $
```
Comment (rvl) [inline lib/byron/exe/cardano-wallet-byron.hs]: Could add a `-- TODO: Error handling` here.

**Review by rvl — APPROVED** [review]
Looks good.

I think this wouldn't handle the case where the node is in a dev cluster with network discrimination disabled and its own genesis. But that probably doesn't matter.

File: lib/byron/exe/cardano-wallet-byron.hs [hunk 2]
```diff
@@ -220,18 +245,68 @@ cmdServe = command "serve" $ info (helper <*> helper' <*> cmd) $ mempty
...
+        someTestnetDiscriminant
+            :: ProtocolMagic
+            -> (SomeNetworkDiscriminant, NodeVersionData)
+        someTestnetDiscriminant pm@(ProtocolMagic n) =
+            case someNatVal (fromIntegral n) of
+                Just (SomeNat proxy) ->
+                    ( SomeNetworkDiscriminant $ mapProxy proxy
+                    , testnetVersionData pm
+                    )
+                _ -> error "networkDiscriminantFlag: impossible: failed to \
```
Comment (paweljakubas) [inline lib/byron/exe/cardano-wallet-byron.hs]: maybe `impossible` was in the earlier version and forgot to erase? `failed to convert ProtocolMagic to SomeNat` is not enough?

File: lib/byron/exe/cardano-wallet-byron.hs [hunk 3]
```diff
@@ -220,18 +245,68 @@ cmdServe = command "serve" $ info (helper <*> helper' <*> cmd) $ mempty
...
+    MainnetConfig
+        :: (SomeNetworkDiscriminant, NodeVersionData)
+        -> NetworkConfiguration
+
+    TestnetConfig
+        :: (SomeNetworkDiscriminant, NodeVersionData)
+        -> FilePath
+        -> NetworkConfiguration
+
+-- | Hand-written as there's no Show instance for 'NodeVersionData'
```
Comment (paweljakubas) [inline lib/byron/exe/cardano-wallet-byron.hs:270]: I wonder why Show and not ToText here (it is needed in `MsgNetworkName Text`) ?

**Review by paweljakubas — APPROVED** [review]
Looks fine (modulo nix voodoo - I am also applying patch from CI rather than producing locally;-)). General comment. Not huge change induced massive number of changes throughout the code. I wonder if we could somehow limit this range...

File: lib/byron/exe/cardano-wallet-byron.hs (hunk 1 above)
Comment (KtorZ) [reply lib/byron/exe/cardano-wallet-byron.hs]: Ho yes! I actually have it on my TODO list and will fix before the PR is merged. Passing a wrong genesis file results in a very ugly error at the moment ...

**Comment by KtorZ** [comment]
> @rvl: network discrimination disabled and its own genesis.

No it wouldn't indeed. There's a strong assumption here that no network discrimination necessarily means mainnet. 

>  @paweljakubas  I wonder if we could somehow limit this range...

Yeah.. I think one of our mistake here has been to specify a network discriminant in EACH integration module containing scenarios, instead of only putting the relevant constraints and only specializing the network in the `integration/Main.hs` (like it is now done for cardano-wallet-byron). As a consequence, we have to go through 50 files instead of 1 in order to change this specialization :(

File: lib/byron/exe/cardano-wallet-byron.hs (hunk 3 above)
Comment (KtorZ) [reply lib/byron/exe/cardano-wallet-byron.hs:270]: That's a good question. I will dig a bit more about why `Show` is needed for `MsgNeworkName`

**Comment by KtorZ** [comment]
> @paweljakubas I wonder if we could somehow limit this range...

Done in 83427e0e3 🥵 ! Removed 32 hard-coded references to network discriminant. Now, the discriminant only appears in a **single** place in the integration scenarios, in `main` in the `integration/Main`.  

```hs
main :: forall t n. (t ~ Jormungandr, n ~ 'Mainnet) => IO ()
main = withUtf8Encoding $ withLogging Nothing Info $ \(_, tr) -> do
    hspec $ do
```

File: lib/byron/exe/cardano-wallet-byron.hs (hunk 1 above)
Comment (KtorZ) [reply lib/byron/exe/cardano-wallet-byron.hs]: Fixed:

```
$ cardano-wallet-byron serve --testnet 1234 --genesis shell.nix  --node-socket .travis.yml 
[cardano-wallet.main:Info:4] [2020-03-16 17:19:22.58 UTC] Running as v2020.3.16 (git revision: 8768ab1a1ab13af5aa9739c47ded3c995121fabe)
[cardano-wallet.main:Info:4] [2020-03-16 17:19:22.58 UTC] Command line: /home/ktorz/.local/bin/cardano-wallet-byron serve --testnet 1234 --genesis shell.nix --node-socket .travis.yml
[cardano-wallet.main:Debug:4] [2020-03-16 17:19:22.58 UTC] ServeArgs {_hostPreference = Host "127.0.0.1", _listen = ListenOnPort 8090, _nodeSocket = ".travis.yml", _networkConfiguration = TestnetConfig "shell.nix", _database = Nothing, _syncTolerance = SyncTolerance 300s, _enableShutdownHandler = False, _logging = LoggingOptions {loggingMinSeverity = Debug, loggingTracers = Tracers {applicationTracer = Const (Just Info), apiServerTracer = Const (Just Info), walletEngineTracer = Const (Just Info), walletDbTracer = Const (Just Info), ntpClientTracer = Const (Just Info)}, loggingTracersDoc = Nothing}}
[cardano-wallet.main:Error:4] [2020-03-16 17:19:22.58 UTC] Failed to parse genesis configuration. You may want to check the filepath given via --genesis and make sure it points to a valid JSON genesis file. Here's (perhaps) some helpful hint: GenesisDataParseError "(line 1, column 1):\nunexpected \"(\"\nexpecting white space or JSON value"
[cardano-wallet.main:Debug:4] [2020-03-16 17:19:22.58 UTC] Logging shutdown.
```

File: lib/byron/exe/cardano-wallet-byron.hs (hunk 2 above)
Comment (KtorZ) [reply lib/byron/exe/cardano-wallet-byron.hs]: fixed.

**Pull request merged**